mqtt/mqtt-broker/src/server.rs

```diff
@@ -225,71 +232,124 @@ where
-                        }
-                    });
-                }
-                Either::Left(_) => {
-                    info!(
-                        "accept loop shutdown. no longer accepting connections on {}",
-                        addr
-                    );
-                    break;
+struct Listener {
+    transport: Transport,
+    authenticator: Arc<(dyn Authenticator<Error = Box<dyn StdError>> + Send + Sync)>,
+    ready: Option<Receiver<()>>,
+    broker_handle: BrokerHandle,
+}
+
+impl Listener {
+    fn new<N>(
+        transport: Transport,
+        authenticator: N,
+        broker_handle: BrokerHandle,
+        ready: Option<Receiver<()>>,
+    ) -> Self
+    where
+        N: Authenticator<Error = Box<dyn StdError>> + Send + Sync + 'static,
+    {
+        Self {
+            transport,
+            authenticator: Arc::new(authenticator),
+            ready,
+            broker_handle,
+        }
+    }
+
+    async fn run<F, P>(self, shutdown_signal: F, make_processor: P) -> Result<(), Error>
+    where
+        F: Future<Output = ()> + Unpin,
+        P: MakeIncomingPacketProcessor
+            + MakeOutgoingPacketProcessor
+            + Clone
+            + Send
+            + Sync
+            + 'static,
+    {
+        let Self {
+            transport,
+            authenticator,
+            ready,
+            broker_handle,
+        } = self;
+
+        let addr = transport.addr();
+
+        let span = info_span!("server", listener=%addr);
+        let inner_span = span.clone();
+
+        async move {
+            let ready = match ready {
+                Some(mut ready) => {
+                    info!("Waiting for broker to be ready to serve requests");
+                    // async block required to consume ready and make a future with
+                    // a 'static lifetime
+                    Either::Left(async move { ready.recv().await })
                 }
-                Either::Right((Some(Err(e)), _)) => {
-                    warn!(
-                        message = "accept loop exiting due to an error",
-                        error =% DetailedErrorValue(&e)
-                    );
-                    break;
+                None => Either::Right(future::ready(Ok(()))),
+            };
+            pin_mut!(ready);
+
+            // wait until broker is ready to serve external clients or a shutdown request
+            match future::select(ready, shutdown_signal).await
+            {
+                Either::Left((Ok(_), mut shutdown_signal)) => {
+                    // start listening incoming connections on given network address
+                    let mut incoming = transport.incoming().await?;
+
+                    let addr = incoming.local_addr()?;
+                    info!("Listening on address {}", addr);
+
+                    loop {
+                        match future::select(&mut shutdown_signal, incoming.next()).await {
```

Comment:
oh, this is interesting `&mut shutdown_signal`.... I always had problems with `future::select` in a loop, b/c it consumed futures in every iteration. Did not know that you can pass mut ref.